agent/logic/agent.py

```diff
@@ -136,8 +142,14 @@ async def __solve(self) -> bool:
         if not solution:
             return retry_if_failed
 
-        await self.__format_solution(solution)
-        return False
+        self.__result_trace.solution = await self.__format_solution(solution)
+
+        retry = self.__prompt_reviser.check_and_revise(
```

Comment:
`check_and_revise` is `async`, but is not awaited here. If this works at all, I would assume `retry` would always evaluate to `True` from the `Coroutine`.